opencensus.go

```diff
@@ -85,6 +86,58 @@ func (c composableRegister) Register(ctx context.Context, cfg Config, vs []*view
 	c.setDefaultSampler(cfg.SampleRate)
 	c.setReportingPeriod(time.Duration(cfg.ReportingPeriod) * time.Second)
 
+	// modify metric tags
+	// ref: https://godoc.org/go.opencensus.io/plugin/ochttp#pkg-variables
+	if cfg.Exporters.Prometheus != nil {
+		for _, view := range vs {
+			// client metrics (method + statuscode tags are enabled by default)
+			if strings.Contains(view.Name, "http/client") {
+				// Host
+				if cfg.Exporters.Prometheus.HostTag {
+					view.TagKeys = appendIfMissing(view.TagKeys, ochttp.KeyClientHost)
+				}
+
+				// Path
+				if cfg.Exporters.Prometheus.PathTag {
+					view.TagKeys = appendIfMissing(view.TagKeys, ochttp.KeyClientPath)
```

Comment:
won't this tag increase the cardinality of the metrics without any control/boundary?

Reply:
Yes - but it's also a very important information for my dashboards as i wouldn't see which endpoints start throwing errors / are under high load in a more aggregated view.

I have 2 ideas what we could do about it:
- overwrite the path before it's tracked, turn `GET /api/product/15` into `GET /api/product/{id}` to get one metric per endpoint (but only if the parameter is at the end of the path, ie. `GET /api/{entity}/{id}` could/should still result in multiple metrics for each `{entity}`)
- do a reset after hitting a specific amount of metrics (ie. 5 x total count of endpoints) to keep memory usage low - prometheus doesn't care about resets/restarts, but you would lose a few seconds of metrics each time this happens.

Im not sure if this has a performance impact too, but if its only a memory usage tradeoff it would be acceptable for me personally even if it results in a lot of metrics (as i want just that). And since it has to be turned on specifically it should be acceptable as it would only be used by those who need it.

Reply:
I think your first suggestion is quite sound. I'd add an intermediate option between no tags and one tag for each url... maybe we could add a https://godoc.org/go.opencensus.io/plugin/ochttp#WithRouteTag wrapper over the generated handler (using the endpoint pattern) at https://github.com/devopsfaith/krakend-opencensus/blob/master/router/mux/endpoint.go#L18 and something similar at https://github.com/devopsfaith/krakend-opencensus/blob/master/router/gin/endpoint.go#L28-L44

I have not benchmarked the impact of the high cardinality over this lib yet, but due previous experiences on the subject I'm quite concerned with it. Believe me, It's not about consuming 500MiB more of RAM. It impacts the overall performance of the gateway (both in RAM, CPU, latency and QoS) and the metrics backend (where your collected data is storage)

On the other hand, I'm not sure if we can reset metrics without messing too much with the internals of the opencensus collectors and exporters

Reply:
I will give it a try, i hope that i have access to both the requested path and the original endpoint (with `.../{paramName}`). If not i would have to iterate over all endpoints in the configuration and find a match every time we get a request which is why i was a bit worried about the first proposal.

Reply:
If the path tag is enabled, the `endpoint pattern` mode will be the default - ie. 1 metric per configured endpoint.

And for a little more flexiblility i made 3 options that can be configured per endpoint via extra options:

- `endpoint` - the default and will directly report the endpoint pattern as path
- `lastparam` - will report the url path and only replace the last section of the url with it's pattern ie. `/api/customers/15` -> `/api/costomers/:customerId` - in case the last part is not a param this will default to reporting the endpoint pattern
- `off` - will report the path without any modification, potentially increases the metric cardinality without limit

I just commited the changes for gin (tested) and mux (not tested, i didn't know how) - and thanks for anwering questions in slack ;)